<a href="https://colab.research.google.com/github/me2527/PulsarStarPrediction/blob/main/01_07_2020_ADITIDWIVEDI_Project6_PulsarStarPrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 6: Predicting A Pulsar Star

### Overview 

Pulsar stars (https://www.youtube.com/watch?v=gjLk_72V9Bw) are a rare type of Neutron stars (https://www.youtube.com/watch?v=hCwDNXKlN8Q) that produce radio emissions detectable on Earth. They are of considerable scientific interest as probes of space-time, the interstellar medium, and states of matter.

You can learn more about the Pulsar and neutron star by watching this video: https://www.youtube.com/watch?v=RrMvUL8HFlM

As pulsars rotate, their emission beam sweeps across the sky, and when this crosses our line of sight, it produces a detectable pattern of broadband radio emission. 

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lightcurve_pulsar.gif'>

As pulsars rotate rapidly, this pattern repeats periodically. Thus pulsar search involves looking for periodic radio signals using large radio telescopes.

Each pulsar produces a slightly different emission pattern, which varies slightly with each rotation. Thus a potential signal detection known as a 'candidate', is averaged over many rotations of the pulsar, as determined by the length of observation. In the absence of additional info, each candidate could potentially describe a real pulsar. However, in practice almost all detections are caused by Radio Frequency Interference (RFI) and noise, making legitimate signals hard to find.

Machine learning tools are now being used to automatically label pulsar candidates to facilitate rapid analysis. The classification algorithms, in particular, are being widely adopted, which treat the candidate datasets as binary classification problems (predict either `0` or `1`). Here, the legitimate pulsar examples are a minority positive class (less in numbers), and the remaining examples are a majority negative class.

The class labels used are `0` (negative class) and `1` (positive class). Hence, **we need to deploy the XGBoost Classifier classification model which can accurately detect the class 1 examples.**


---

### Attribute Information

Each candidate is described by 8 continuous variables and a single class variable. The first four are simple statistics obtained from the integrated pulse profile (folded profile). This is an array of continuous variables that describe a longitude-resolved version of the signal that has been averaged in both time and frequency. The remaining four variables are similarly obtained from the DM-SNR curve. **You do not have to worry about what they actually mean.** These 8 variables are summarised below:

1. The mean of the integrated profile.

2. The standard deviation of the integrated profile.

3. Excess kurtosis of the integrated profile.

4. The skewness of the integrated profile.

5. The mean of the DM-SNR curve.

6. The standard deviation of the DM-SNR curve.

7. Excess kurtosis of the DM-SNR curve.

8. The skewness of the DM-SNR curve.

Source: https://archive.ics.uci.edu/ml/datasets/HTRU2

**Courtesy**

Dr Robert Lyon

University of Manchester

School of Physics and Astronomy

Alan Turing Building

Manchester M13 9PL

United Kingdom

robert.lyon@manchester.ac.uk

#### Dataset Links

You can get the datasets from the following links:

1. Train dataset: https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-5/pulsar-star-prediction-train.csv

2. Test dataset: https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-5/pulsar-star-prediction-test.csv


---

### Project Requirements

1. Create a pandas DataFrame for both the train and test datasets.

2. Display the first five rows of both the training and test DataFrames.

3. Display the last five rows of both the training and test DataFrames.

4. Find the number of rows and columns in both the train and test DataFrames.

5. Check for the missing values in both the train and test DataFrames.

6. Count the number of `0` and `1` classes in the training dataset.

7. Separate the feature variables, i.e, `x_train` and `x_test` from both the DataFrames.

8. Separate the target variable, i.e, `y_train` and `y_test` from both the DataFrames.

9. Apply the `XGBClassifier` machine learning model to predict the `0` and `1` classes in the test dataset, i.e, `x_test`.

10. Print the confusion matrix and the classification report to evaluate your prediction model. Also, based on the confusion matrix, precision, recall and f1-score values, report whether the prediction model deployed by you is making accurate predictions or not.


---

#### 1. Create The DataFrames

Create the Pandas DataFrames for both the training and test datasets.

In [ ]:
# Create the DataFrames for both the train and test datasets and store them in the 'train_df' and 'test_df' variables respectively.
import pandas as pd
train_df=pd.read_csv("https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-5/pulsar-star-prediction-train.csv")
test_df=pd.read_csv("https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-5/pulsar-star-prediction-test.csv")

---

#### 2. Display First Five Rows

Display the first five rows of both the `train_df` and `test_df` DataFrames.

In [ ]:
# Display the first 5 rows of the 'train_df' DataFrame.
train_df.head()

target_class  ...   Skewness of the DM-SNR curve
0             0  ...                      62.383270
1             0  ...                     173.592172
2             1  ...                      21.987882
3             0  ...                      98.584122
4             0  ...                     196.522501

[5 rows x 9 columns]

In [ ]:
# Display the first 5 rows of the 'test_df' DataFrame.
test_df.head()

target_class  ...   Skewness of the DM-SNR curve
0             0  ...                      78.813405
1             1  ...                      55.878791
2             0  ...                     105.134941
3             1  ...                      -1.117904
4             0  ...                      21.708268

[5 rows x 9 columns]

---

#### 3. Display Last Five Rows

Display the last five rows of both the `train_df` and `test_df` DataFrames.

In [ ]:
# Display the last 5 rows of the 'train_df' DataFrame.
train_df.tail()

target_class  ...   Skewness of the DM-SNR curve
11986             0  ...                      21.725143
11987             0  ...                      34.357283
11988             0  ...                     312.204325
11989             0  ...                      62.334018
11990             0  ...                     329.548016

[5 rows x 9 columns]

In [ ]:
# Display the last 5 rows of the 'test_df' DataFrame.
test_df.tail()

target_class  ...   Skewness of the DM-SNR curve
5902             0  ...                     111.061145
5903             0  ...                      80.304937
5904             0  ...                     167.163044
5905             0  ...                     431.240745
5906             0  ...                      30.444595

[5 rows x 9 columns]

---

#### 4. Rows & Columns In Train DataFrame

Find the number of rows and columns in both the `train_df` and `test_df` DataFrames.

In [ ]:
# Print the number of rows and columns in both the DataFrames.
print("Rows and Columns in training datasets is ",train_df.shape)
print("Rows and Columns in testing datasets is ",test_df.shape)

Rows and Columns in training datasets is  (11991, 9)
Rows and Columns in testing datasets is  (5907, 9)


---

#### 5. Check For Missing Values 

Check whether any of the columns in both the `train_df` and `test_df` DataFrames has any missing value.

In [ ]:
# Check for the missing values in the 'train_df' DataFrame.
num_missing=0
for column in train_df.columns:
  for item in train_df[column].isnull():
    if item==False:
      num_missing+=1
print("Missing Values for false values:",num_missing)

num_miss=0
for column in train_df.columns:
  for item in train_df[column].isnull():
    if item==True:
      num_miss+=1
print("Missing Values for true values:",num_miss)

Missing Values for false values: 107919
Missing Values for true values: 0


In [ ]:
# Check for the missing values in the 'test_df' DataFrame.
missing_num=0
for column in test_df.columns:
  for item in test_df[column].isnull():
    if item==False:
      missing_num+=1
print("Missing Values for false values:",missing_num)

miss_num=0
for column in test_df.columns:
  for item in test_df[column].isnull():
    if item==True:
      num_miss+=1
print("Missing Values for true values:",miss_num)

Missing Values for false values: 53163
Missing Values for true values: 0


---

#### 6. Count The `0` & `1` Classes

Find out the number of `0` and `1` values in both the `train_df` and `test_df` DataFrames.

In [ ]:
# Print the count of the '0' and '1' classes in the 'train_df' DataFrame.
print(train_df['target_class'].value_counts(0))
print(train_df['target_class'].value_counts(1))

0    10878
1     1113
Name: target_class, dtype: int64
0    0.90718
1    0.09282
Name: target_class, dtype: float64


In [ ]:
# Print the count of the '0' and '1' classes in the 'test_df' DataFrame.
print(test_df['target_class'].value_counts(0))
print(test_df['target_class'].value_counts(1))

0    5381
1     526
Name: target_class, dtype: int64
0    0.910953
1    0.089047
Name: target_class, dtype: float64


---

#### 7. Feature Variables Extraction

Get the feature variables, i.e., `x_train` and `x_test` from both the `train_df` and `test_df` DataFrames respectively. Then, display the first 5 rows of `x_train` and `x_test` DataFrames.

In [ ]:
# Get the feature variables from the 'train_df' DataFrame.
x_train=train_df.iloc[:,1:]
# Display the first 5 rows of the 'x_train' DataFrame.
x_train.head()

Mean of the integrated profile  ...   Skewness of the DM-SNR curve
0                       111.109375  ...                      62.383270
1                       151.945312  ...                     173.592172
2                        52.335938  ...                      21.987882
3                       121.562500  ...                      98.584122
4                       133.664062  ...                     196.522501

[5 rows x 8 columns]

In [ ]:
# Get the feature variables from the 'test_df' DataFrame.
x_test=test_df.iloc[:,1:]
# Display the first 5 rows of the 'x_test' DataFrame.
x_test.head()

Mean of the integrated profile  ...   Skewness of the DM-SNR curve
0                       116.906250  ...                      78.813405
1                        75.585938  ...                      55.878791
2                       103.273438  ...                     105.134941
3                       101.078125  ...                      -1.117904
4                       113.226562  ...                      21.708268

[5 rows x 8 columns]

---

#### 8. Target Variable Extraction

Get the target variables, i.e., `y_train` and `y_test` from both the `train_df` and `test_df` DataFrames respectively. Then, display the first 5 rows of `y_train` and `y_test` Pandas series.

In [ ]:
# Get the target variable from the 'train_df' DataFrame.
y_train=train_df["target_class"]
# Display the first 5 rows of the 'y_train' Pandas series.
y_train.head()

0    0
1    0
2    1
3    0
4    0
Name: target_class, dtype: int64

In [ ]:
# Get the target variable from the 'test_df' DataFrame.
y_test=test_df["target_class"]
# Display the first 5 rows of the 'y_test' Pandas series.
y_test.head()

0    0
1    1
2    0
3    1
4    0
Name: target_class, dtype: int64

--- 

#### 9. Building An XGBoost Classifier Model

In [ ]:
# Build A XGBoost Classifier model
# Import the xgboost module.
import xgboost as xgb
# Import the confusion_matrix and classification_report modules.
from sklearn.metrics import confusion_matrix, classification_report

In [ ]:
# Predict the target variable based on the feature variables of the test dataframe.
model=xgb.XGBClassifier()
model.fit(x_train,y_train)
y_pred=model.predict(x_test)
print(y_pred)


[0 1 0 ... 0 0 0]


---

#### 10. Confusion Matrix & Classification Report

Print the confusion matrix and classification report to evaluate the model. Interpret and report the results obtained from the confusion matrix and the classification report.

In [ ]:
# Print the confusion matrix to see the number of TN, FN, TP and FP.
print(confusion_matrix(y_test,y_pred))

[[5348   33]
 [  82  444]]


In [ ]:
# Print the precision, recall and f1-score values for both the '0' and '1' classes.
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.98      0.99      0.99      5381
           1       0.93      0.84      0.89       526

    accuracy                           0.98      5907
   macro avg       0.96      0.92      0.94      5907
weighted avg       0.98      0.98      0.98      5907



**Write your interpretation of the results here.**

- Interpretation: The precison, recall & f1-score values are reported as 0.93, 0.84,0.89 for class 1 respectively and for class 0 model reported precision, recall, f1-score as 0.98,.99,0.99 respectively.
So, our prediction model is quite close to satisfactory facts. Hence,we can use this deployed model further for testing a dataset.


---

### How To Solve & Submit The Project

Follow the steps described below to solve the project and then submit it.

1. Click on the link provided below to open the Colab file for this project.
   
   https://colab.research.google.com/drive/1jK_kbtQCGjb5mIYLeXQeKIqjiHI3VGhg

2. Create the duplicate copy of the Colab file. Here are the steps to create the duplicate copy:

    - Click on the **File** menu. A new drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/0_file_menu.png' width=500>

    - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/1_create_colab_duplicate_copy.png' width=500>

     - After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_Project6** format. 

3. Now, write your code in the prescribed code cells.

4. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/2_share_button.png' width=500>

5. In the dialog box, click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-1/1_copy_link.png' width=500>

6. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_Project6**) of the notebook will get copied which you can paste in the submission box on the WhiteHat Jr project submission interface.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-1/2_copy_link_confirmation.png' width=500>

---